### PR TITLE
[WIP][TensorExpr] Add a tensor expressions tutorial.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45453 [WIP][TensorExpr] Add a tensor expressions tutorial.**
* #45520 New work wihtout tutorial update

Differential Revision: [D23972887](https://our.internmc.facebook.com/intern/diff/D23972887)

### CODE REVIEW

File: test/cpp/tensorexpr/tutorial.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    // An ultimate purpose of tensor expressions is to optimize tensor
+    // computations, and in order to represent accesses to tensors data, there
+    // is a special kind of expression - a load.
+    // To construct a load we need two pieces: the base and the indices. The
+    // base of a load is a Buf expression, which could be thought of as a
+    // placeholder similar to Var, but with dimensions info.
+    //
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
```
Comment (Lilyjjo) [inline test/cpp/tensorexpr/tutorial.cpp:66]: Is a BufHandle a 'Buf expression'?

File: test/cpp/tensorexpr/tutorial.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    Placeholder A_buf(A);
```
Comment (Lilyjjo) [inline test/cpp/tensorexpr/tutorial.cpp:67]: Is there a quick way to explain why a Placeholder is needed to wrap the BufHandle?

File: test/cpp/tensorexpr/tutorial.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply test/cpp/tensorexpr/tutorial.cpp:66]: Yes, `Buf` is an `Expr` too. And similarly to how `Expr` has `ExprHandle`, `Buf` has `BufHandle`. FWIW, `Var` is also an `Expr`, and also has `VarHandle`.

"Buf" might be not the best name for this and there are chances it gets renamed soon.

File: test/cpp/tensorexpr/tutorial.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply test/cpp/tensorexpr/tutorial.cpp:67]: The primary reason is to expose a separate class in users' API while hiding internal classes (the definition of users is vague here, but you could think of several layers of details in API: the highest level of abstraction is when you use `Compute` API and never construct expressions manually). In this sense, `Tensor` and `Placeholder` are what we expose and `Buf` is what they are built upon internally.

But technically `Placeholder` doesn't add anything to what `Buf` already has.

**Pull request merged**